`quarkus.hibernate-orm.enabled = false` doesn't work when using Panache

### Describe the bug

If we try to set `quarkus.hibernate-orm-enabled = false` in an application that uses Panache, we get this error when building:

```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.13.2.Final:build (default) on project managed-server-ui-cli: Failed to build quarkus application: io.quarkus.builder.ChainBuildException: No producers for required item class io.quarkus.hibernate.orm.deployment.JpaModelBuildItem -> [Help 1]
```

That's probably caused by `PanacheJpaCommonResourceProcessor` assuming that `JpaModelBuildItem` is always present; it should use `Optional<JpaModelBuildItem>` instead. Or it should just disable itself when Hibernate ORM is disabled.

And we should have tests for that...

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/panache/hibernate-orm-panache-common/deployment/src/main/java/io/quarkus/hibernate/orm/panache/common/deployment/PanacheJpaCommonResourceProcessor.java
+++ b/extensions/panache/hibernate-orm-panache-common/deployment/src/main/java/io/quarkus/hibernate/orm/panache/common/deployment/PanacheJpaCommonResourceProcessor.java
@@ -17,13 +17,16 @@ import org.jboss.jandex.Type;
 
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.BuildSteps;
 import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.util.JandexUtil;
+import io.quarkus.hibernate.orm.deployment.HibernateOrmEnabled;
 import io.quarkus.hibernate.orm.deployment.JpaModelBuildItem;
 import io.quarkus.hibernate.orm.panache.common.runtime.PanacheHibernateRecorder;
 
+@BuildSteps(onlyIf = HibernateOrmEnabled.class)
 public final class PanacheJpaCommonResourceProcessor {
 
     private static final DotName DOTNAME_NAMED_QUERY = DotName.createSimple(NamedQuery.class.getName());

--- a/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/MyEntity.kt
+++ b/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/MyEntity.kt
@@ -1,0 +1,13 @@
+package io.quarkus.hibernate.orm.panache.kotlin.deployment.test
+
+import io.quarkus.hibernate.orm.panache.kotlin.PanacheCompanion
+import io.quarkus.hibernate.orm.panache.kotlin.PanacheEntity
+import javax.persistence.Entity
+
+@Entity
+class MyEntity : PanacheEntity() {
+    companion object: PanacheCompanion<MyEntity> {
+    }
+
+    lateinit var name: String
+}

--- a/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/config/ConfigEnabledFalseTest.kt
+++ b/extensions/panache/hibernate-orm-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/orm/panache/kotlin/deployment/test/config/ConfigEnabledFalseTest.kt
@@ -1,0 +1,29 @@
+package io.quarkus.hibernate.orm.panache.kotlin.deployment.test.config
+
+import io.quarkus.arc.Arc
+import io.quarkus.hibernate.orm.panache.kotlin.deployment.test.MyEntity
+import io.quarkus.test.QuarkusUnitTest
+import org.jboss.shrinkwrap.api.spec.JavaArchive
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+import javax.persistence.EntityManagerFactory
+
+class ConfigEnabledFalseTest {
+    companion object {
+        @RegisterExtension
+        val config = QuarkusUnitTest()
+            .withApplicationRoot { jar: JavaArchive -> jar.addClass(MyEntity::class.java) }
+            .withConfigurationResource("application-test.properties")
+            // We shouldn't get any build error caused by Panache consuming build items that are not produced
+            // See https://github.com/quarkusio/quarkus/issues/28842
+            .overrideConfigKey("quarkus.hibernate-orm.enabled", "false")
+    }
+
+    @Test
+    fun startsWithoutError() {
+        // Quarkus started without problem, even though the Panache extension is present.
+        // Just check that Hibernate ORM is disabled.
+        Assertions.assertNull(Arc.container().instance(EntityManagerFactory::class.java).get())
+    }
+}

--- a/extensions/panache/hibernate-orm-panache/deployment/pom.xml
+++ b/extensions/panache/hibernate-orm-panache/deployment/pom.xml
@@ -68,6 +68,11 @@
             <artifactId>quarkus-resteasy-deployment</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/config/ConfigEnabledFalseTest.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/config/ConfigEnabledFalseTest.java
@@ -1,0 +1,31 @@
+package io.quarkus.hibernate.orm.panache.deployment.test.config;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import javax.persistence.EntityManagerFactory;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.hibernate.orm.panache.deployment.test.MyEntity;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ConfigEnabledFalseTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot(jar -> jar.addClass(MyEntity.class))
+            .withConfigurationResource("application-test.properties")
+            // We shouldn't get any build error caused by Panache consuming build items that are not produced
+            // See https://github.com/quarkusio/quarkus/issues/28842
+            .overrideConfigKey("quarkus.hibernate-orm.enabled", "false");
+
+    @Test
+    public void startsWithoutError() {
+        // Quarkus started without problem, even though the Panache extension is present.
+        // Just check that Hibernate ORM is disabled.
+        assertThat(Arc.container().instance(EntityManagerFactory.class).get())
+                .isNull();
+    }
+}

--- a/extensions/panache/hibernate-reactive-panache-common/deployment/src/main/java/io/quarkus/hibernate/reactive/panache/deployment/PanacheJpaCommonResourceProcessor.java
+++ b/extensions/panache/hibernate-reactive-panache-common/deployment/src/main/java/io/quarkus/hibernate/reactive/panache/deployment/PanacheJpaCommonResourceProcessor.java
@@ -23,16 +23,19 @@ import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
 import io.quarkus.deployment.IsTest;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.BuildSteps;
 import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.util.JandexUtil;
 import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.hibernate.orm.deployment.HibernateOrmEnabled;
 import io.quarkus.hibernate.orm.deployment.JpaModelBuildItem;
 import io.quarkus.hibernate.reactive.panache.common.runtime.PanacheHibernateRecorder;
 import io.quarkus.hibernate.reactive.panache.common.runtime.ReactiveTransactionalInterceptor;
 import io.quarkus.hibernate.reactive.panache.common.runtime.TestReactiveTransactionalInterceptor;
 
+@BuildSteps(onlyIf = HibernateOrmEnabled.class)
 public final class PanacheJpaCommonResourceProcessor {
 
     private static final DotName DOTNAME_NAMED_QUERY = DotName.createSimple(NamedQuery.class.getName());

--- a/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/deployment/test/MyEntity.kt
+++ b/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/deployment/test/MyEntity.kt
@@ -1,0 +1,13 @@
+package io.quarkus.hibernate.reactive.panache.kotlin.deployment.test
+
+import io.quarkus.hibernate.reactive.panache.kotlin.PanacheCompanion
+import io.quarkus.hibernate.reactive.panache.kotlin.PanacheEntity
+import javax.persistence.Entity
+
+@Entity
+class MyEntity : PanacheEntity() {
+    companion object: PanacheCompanion<MyEntity> {
+    }
+
+    lateinit var name: String
+}

--- a/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/deployment/test/config/ConfigEnabledFalseTest.kt
+++ b/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/deployment/test/config/ConfigEnabledFalseTest.kt
@@ -1,0 +1,30 @@
+package io.quarkus.hibernate.reactive.panache.kotlin.deployment.test.config
+
+import io.quarkus.arc.Arc
+import io.quarkus.hibernate.reactive.panache.kotlin.deployment.test.MyEntity
+import io.quarkus.test.QuarkusUnitTest
+import org.hibernate.reactive.mutiny.Mutiny
+import org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl
+import org.jboss.shrinkwrap.api.spec.JavaArchive
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+
+class ConfigEnabledFalseTest {
+    companion object {
+        @RegisterExtension
+        val config = QuarkusUnitTest()
+            .withApplicationRoot { jar: JavaArchive -> jar.addClass(MyEntity::class.java) }
+            .withConfigurationResource("application.properties")
+            // We shouldn't get any build error caused by Panache consuming build items that are not produced
+            // See https://github.com/quarkusio/quarkus/issues/28842
+            .overrideConfigKey("quarkus.hibernate-orm.enabled", "false")
+    }
+
+    @Test
+    fun startsWithoutError() {
+        // Quarkus started without problem, even though the Panache extension is present.
+        // Just check that Hibernate Reactive is disabled.
+        Assertions.assertNull(Arc.container().instance(Mutiny.SessionFactory::class.java).get())
+    }
+}

--- a/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/resources/application.properties
+++ b/extensions/panache/hibernate-reactive-panache-kotlin/deployment/src/test/resources/application.properties
@@ -1,0 +1,7 @@
+quarkus.datasource.db-kind=postgresql
+quarkus.datasource.username=hibernate_orm_test
+quarkus.datasource.password=hibernate_orm_test
+quarkus.datasource.reactive.url=${postgres.reactive.url}
+quarkus.datasource.devservices.enabled=false
+
+quarkus.hibernate-orm.database.generation=drop-and-create

--- a/extensions/panache/hibernate-reactive-panache/deployment/pom.xml
+++ b/extensions/panache/hibernate-reactive-panache/deployment/pom.xml
@@ -55,6 +55,11 @@
             <artifactId>rest-assured</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/panache/hibernate-reactive-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/panache/test/config/ConfigEnabledFalseTest.java
+++ b/extensions/panache/hibernate-reactive-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/panache/test/config/ConfigEnabledFalseTest.java
@@ -1,0 +1,30 @@
+package io.quarkus.hibernate.reactive.panache.test.config;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.hibernate.reactive.mutiny.Mutiny;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.hibernate.reactive.panache.test.MyEntity;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ConfigEnabledFalseTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot(jar -> jar.addClass(MyEntity.class))
+            .withConfigurationResource("application.properties")
+            // We shouldn't get any build error caused by Panache consuming build items that are not produced
+            // See https://github.com/quarkusio/quarkus/issues/28842
+            .overrideConfigKey("quarkus.hibernate-orm.enabled", "false");
+
+    @Test
+    public void startsWithoutError() {
+        // Quarkus started without problem, even though the Panache extension is present.
+        // Just check that Hibernate ORM is disabled.
+        assertThat(Arc.container().instance(Mutiny.SessionFactory.class).get())
+                .isNull();
+    }
+}

